crates/marine-build-rs-generator/src/lib.rs

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2021 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#![doc(html_root_url = "https://docs.rs/marine-build-rs-generator/0.4.0")]
+#![deny(
+    dead_code,
+    nonstandard_style,
+    unused_imports,
+    unused_mut,
+    unused_variables,
+    unused_unsafe,
+    unreachable_patterns
+)]
+#![warn(rust_2018_idioms)]
+#![recursion_limit = "1024"]
+
+use marine_test_macro_impl::generate_marine_test_env_impl;
+use std::collections::HashMap;
+
+pub use marine_test_macro_impl::ServiceDescription;
+use std::path::{PathBuf, Path};
+use std::{fs, env};
+
+pub fn generate_marine_test_env(
+    services: &[(String, ServiceDescription)],
+    filename: &str,
+    build_rs_file_path: &str,
+) {
+    let services = services
+        .iter()
+        .cloned()
+        .collect::<HashMap<String, ServiceDescription>>();
+
+    let mut build_rs_file_path = PathBuf::from(build_rs_file_path);
+    let _ = build_rs_file_path.pop();
+
+    match generate_marine_test_env_impl(services, &build_rs_file_path) {
+        Ok(stream) => {
+            let out_dir = env::var_os("OUT_DIR")
+                .expect("cannot write marine_test_env: OUT_DIR env var must be set");
+            let dest_path = Path::new(&out_dir).join(filename);
+
+            let result = fs::write(dest_path, stream.to_string());
+            if let Err(e) = result {
+                std::panic::panic_any(format!(
+                    "cannot write marine_test_env on disk: {}",
+                    e.to_string()
+                ));
+            }
+        }
+        Err(error) => std::panic::panic_any(format!(
+            "marine_test_env generation error: {}",
+            error.to_string()
```

Comment:
why do you need `to_string()` here?

Reply:
IDE highlighted it as an error, saying that TestGenerationError does not implement Display. But, in fact, it works well without to_string